plugins/utils/which-key.nix

```diff
@@ -7,23 +7,94 @@
   ...
 }:
 with lib;
-{
-  options.plugins.which-key = {
-    enable = mkEnableOption "which-key.nvim, a plugin that popup with possible key bindings of the command you started typing";
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "which-key";
+  originalName = "which-key.nvim";
+  defaultPackage = pkgs.vimPlugins.which-key-nvim;
 
-    package = helpers.mkPluginPackageOption "which-key-nvim" pkgs.vimPlugins.which-key-nvim;
+  maintainers = [ lib.maintainers.khaneliman ];
 
-    registrations = mkOption {
-      type = with types; attrsOf anything;
-      description = ''
-        This option is deprecated, use `settings.spec` instead.
+  # TODO: introduced 2024-08-05: remove after 24.11
+  optionsRenamedToSettings = [
+    "hidden"
+    "icons"
+    "ignoreMissing"
+    "keyLabels"
+    "layout"
+    "motions"
+    "operators"
+    "plugins"
+    "popupMappings"
+    "showHelp"
+    "showKeys"
+    "triggers"
+    "triggersBlackList"
+    "triggersNoWait"
+  ];
 
-        Note: the keymap format has changed in v3 of which-key.
-      '';
-      visible = false;
-    };
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "which-key"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (lib.mkRenamedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "disable"
+            "buftypes"
+          ]
+        )
+        (
+          settingsPath
+          ++ [
+            "disable"
+            "bt"
+          ]
+        )
+      )
+      (lib.mkRenamedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "disable"
+            "filetypes"
+          ]
+        )
+        (
+          settingsPath
+          ++ [
+            "disable"
+            "ft"
+          ]
+        )
+      )
+    ];
 
-    settings.spec = helpers.defaultNullOpts.mkListOf' {
+  settingsOptions = {
+    preset = helpers.defaultNullOpts.mkNullable (types.either types.bool (
+      types.enum [
+        "classic"
+        "modern"
+        "helix"
+      ]
+    )) "classic" "Preset style for WhichKey. Set to false to disable.";
+
+    delay = helpers.defaultNullOpts.mkInt (helpers.mkRaw ''
+      function(ctx)
+        return ctx.plugin and 0 or 200
+      end'') "Delay before showing the popup. Can be a number or a function that returns a number.";
+
+    filter = helpers.defaultNullOpts.mkLuaFn (helpers.mkRaw ''
+      function(mapping)
+        return true
+      end'') "Filter used to exclude mappings";
```

Comment:
These may be more readable if declared using the `mk'` variants?

Although, we're (slowly) coming around to the idea of declaring fewer `settingsOptions` and relying more on the freeformType.

So it may make more sense _not_ to include some of these lesser used options and instead point users to the upstream docs.